### PR TITLE
Profiles: add QuickMenu

### PR DESCRIPTION
**Pull request by hius07**
Description and sreenshots in https://github.com/koreader/koreader/issues/8701#issuecomment-1249608740. Closes https://github.com/koreader/koreader/issues/8701.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9526)

### CODE REVIEW

File: frontend/ui/widget/button.lua [hunk 1]
```diff
@@ -78,8 +80,11 @@ function Button:init()
         self.padding_v = self.padding
     end
 
+    local is_left = self.align == "left"
+    local r_margin = is_left and (2 * Size.padding.large) or 0
```
Comment (poire-z) [inline frontend/ui/widget/button.lua:84]: You can be more verbose: `is_left_aligned` , `right_margin`.

File: plugins/profiles.koplugin/main.lua [hunk 2]
```diff
@@ -41,6 +43,25 @@ function Profiles:onFlushSettings()
...
 
+local function dispatcherRegisterProfile(name, unregister)
+    if unregister then
+        Dispatcher:removeAction("profile_exec_"..name)
+        Dispatcher:removeAction("profile_menu_"..name)
+    else
+        Dispatcher:registerAction("profile_exec_"..name,
+            {category="none", event="ProfileExecute", arg=name, title=T(_("Profile execute: %1"), name), general=true})
+        Dispatcher:registerAction("profile_menu_"..name,
+            {category="none", event="ProfileShowMenu", arg=name, title=T(_("Profile show menu: %1"), name), general=true})
```
Comment (poire-z) [inline plugins/profiles.koplugin/main.lua:54]:
```
Profile execute: %1
Profile show menu: %1
```
These "sounds" a bit programmatically, and not super nice English.
No real alternative idea though :/

Also, "menu" is our internal name for such lists (for a user, Menu is something at the top, that pops down :), dunno if a better name could be used, like "popup", "popup list".
QuickMenu (below) sounds a bit better as a nickname for that thing.

File: plugins/profiles.koplugin/main.lua [hunk 3]
```diff
@@ -57,136 +78,320 @@ function Profiles:getSubMenuItems()
...
+                            touchmenu_instance:updateItems()
+                        end,
+                    })
+                end,
             },
         }
-        Dispatcher:addSubMenu(self, sub_items, self.data, k)
+        local edit_actions_sub_items = {}
+        Dispatcher:addSubMenu(self, edit_actions_sub_items, self.data, k)
+        table.insert(sub_items, 5, {
```
Comment (poire-z) [inline plugins/profiles.koplugin/main.lua]: (I'm not having a great look at the whole context, but any reason to table.insert it here at the specific 5th position, vs. doing it inline like all other menu items ?)

File: plugins/profiles.koplugin/main.lua (hunk 2 above)
Comment (Frenzie) [reply plugins/profiles.koplugin/main.lua:54]: What exactly do they mean?

Comment (poire-z) [reply plugins/profiles.koplugin/main.lua:54]:
```
Execute profile.
Show profile as a menu/popup.
```
(I guess the current naming is to have them both started with `Profile...` to make it obvious they come from the same Profile plugin.

Comment (Frenzie) [reply plugins/profiles.koplugin/main.lua:54]: I can see the attraction, but I think it'd be better to simply name them that.

File: frontend/ui/widget/buttondialogtitle.lua [hunk 4]
```diff
@@ -32,6 +34,14 @@ local ButtonDialogTitle = InputContainer:new{
 }
 
 function ButtonDialogTitle:init()
+    self.screen_width = Screen:getWidth()
+    self.screen_height = Screen:getHeight()
+    if not self.width then
+        if not self.width_factor then
+            self.width_factor = 0.9 -- default if no width speficied
```
Comment (NiLuJe) [inline frontend/ui/widget/buttondialogtitle.lua:41]: nit: specified

File: frontend/ui/widget/buttondialogtitle.lua [hunk 5]
```diff
@@ -65,13 +75,14 @@ function ButtonDialogTitle:init()
                         bordersize = 0,
                         TextBoxWidget:new{
                             text = self.title,
-                            width = math.floor(math.min(Screen:getWidth(), Screen:getHeight()) * 0.8),
+                            width = self.width * 0.9,
```
Comment (NiLuJe) [inline frontend/ui/widget/buttondialogtitle.lua:78]: nit: we still need the truncation here ;).

File: plugins/profiles.koplugin/main.lua (hunk 2 above)
Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:54]: The first one could kinda work as `Profile execution` but I don't have anything better for the second than `Profile menu`, and both are arguably worse than the verbose approach ;).

File: plugins/profiles.koplugin/main.lua [hunk 6]
```diff
@@ -57,136 +78,320 @@ function Profiles:getSubMenuItems()
...
-    self.data[name] = nil
-    self.updated = true
-    self:deleteAutostartProfile(name)
+function Profiles:onProfileShowMenu(name)
+    if UIManager:getTopWidget() == name then return end
+    local profile = self.data[name]
+    local actions_list = self:getActionsList(name)
+    local quickmenu
+    local buttons = {}
+    for i = 1, #actions_list do
```
Comment (NiLuJe) [inline plugins/profiles.koplugin/main.lua:244]: `ipairs`?

File: plugins/profiles.koplugin/main.lua [hunk 7]
```diff
@@ -57,136 +78,320 @@ function Profiles:getSubMenuItems()
...
+    sort_widget = SortWidget:new{
+        title = _("Sort actions"),
+        item_table = actions_list,
+        callback = function()
+            if profile.settings then
+                self.data[name].settings.actions_order = {}
+            else
+                self.data[name].settings = {["actions_order"] = {}}
+            end
+            for i = 1, #sort_widget.item_table do
```
Comment (NiLuJe) [inline plugins/profiles.koplugin/main.lua:284]: `ipairs`?

File: plugins/profiles.koplugin/main.lua (hunk 2 above)
Comment (hius07) [reply plugins/profiles.koplugin/main.lua:54]: We need something short, to show more profile name. `Profile exec:`, `Profile show:`.

Comment (Frenzie) [reply plugins/profiles.koplugin/main.lua:54]: Maybe some unicode emoji icon type thing?

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:54]: Hmm, interesting idea. Nothing clear jumps to mind right now, though :/.

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:54]: Yeah, a checkmarked listcame to mind, not sure it's super clear though.

Did not think of the "shortcut/launcher" approach.

Oh, also just realized you meant to replace the *verb* by an icon, my train of thought was to replace "profile" itself, which was more annoying ;p.

TL;DR: I like the first one ;p.

Comment (poire-z) [reply plugins/profiles.koplugin/main.lua:54]: I like image for the quickmenu ! Not so sure about the ones for exec :)
Remember we have most of https://www.nerdfonts.com/ - a bit tricky to get their codepoint though (some shift is needed, needs to look at https://github.com/koreader/koreader-fonts/pull/8#issuecomment-554145620 and some trials).

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:54]: Yeah, that could work. Plasma uses a similar approach for its "Open with" icon.

(I checked because I wanted to check what was used for the "Execute" entry on a shell script, but that's just the "terminal" angled bracket, which would possibly be more obscure for non-Linux/macOS users ;)).

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:54]: I mean, maybe don't elide "Profile" on submenus that don't directly come from a Profile-related thingy, though?

Comment (NiLuJe) [reply plugins/profiles.koplugin/main.lua:54]: > a bit tricky to get their codepoint though

FWIW, my usual approach is to get the name from the nerdfonts website, then look it up inside our own font w/ fontforge ;).

Comment (hius07) [reply plugins/profiles.koplugin/main.lua:54]: I would keep the "Profile" word, it helps in the long list of dispatcher actions.

**Review by poire-z — APPROVED** [review]

**Review by Frenzie — APPROVED** [review]

File: frontend/ui/widget/buttondialogtitle.lua (hunk 5 above)
Comment (NiLuJe) [inline frontend/ui/widget/buttondialogtitle.lua:78]: Still needs a `math.floor` ;).

**Comment by yparitcher** [comment]
You guys snuck this in quickly :)
will review/comment when i have some time

**Comment by yparitcher** [comment]
A few quick comments.
I was planning on making `Dispatcher` sortable but never got around to it, Thanks @hius07 for the implementation. I will probably steal it and integrate it directly into dispatcher.

I will state again for the record :) that i am against registering profiles in dispatcher as this can just lead to loops and generally should not be needed.
(I might also adapt `onProfileShowMenu` into `Dispatcher`)

**Comment by hius07** [comment]
I thought of removing automatic registration of the profiles. It would be done manually with a long-press on Execute or Show as QuickMenu (the check boxes should be added).
This would enforce the user to think about his use cases and would reduce the risk of loops.

I like the QuickMenus very much )) They are easy to use and save much time.

File: plugins/profiles.koplugin/main.lua (hunk 2 above)
Comment (mergen3107) [reply plugins/profiles.koplugin/main.lua]: I think we should add these icons next to the "Execute" and "Show as QuickMenu" so the user will recognize them immediately.

**Comment by mergen3107** [comment]
I tested this quickly and have some thoughts:
1) The ease of access to those profiles got improved tremendously! Thank you :) As a further improvement, can there be a new item in the Dispatcher called "Show list of profiles"? So that with one gesture the user can summon the whole list of profiles? With long-press activating them or showing their menu (to be consistent with the Profiles top menu) maybe.

2) There is option called "Copy". I think this better be named "Duplicate" since "Copy" implies (at least in my UX head) that it is going to be "Pasted" somewhere. However, there is no other place where this profile structure can be pasted.

3) When editing the actions of a profile, the same Dispatcher item list is used. This means that, for example, I can add the action "Execute Profile 2" in Profile 1 and "Execute Profile 1" in Profile 2. So this is going to be a loop. Haven't' tested this myself though. Shouldn't we better remove those profile-specific stuff from the profile actions editing list? 

4) The old small nitpicks from my original https://github.com/koreader/koreader/issues/9391#issuecomment-1204231095 are still there.

Again, thank you for your PR! KOReader rocks!

**Comment by hius07** [comment]
> add these icons next to the "Execute" and "Show as QuickMenu" so the user will recognize them immediately.

Maybe, no strong opinion.






> new item in the Dispatcher called "Show list of profiles"

Maybe. Short-press invokes profile QuickMenu, long-press executes the profile.

**Comment by hius07** [comment]
> * When selecting some settings, like `Font`, `Font weight`, the menu with choices stays. However, since the these parameters are not applied right now, there is no point of staying here. So I think it is better to leave the menu once the choice is made. These are not the parameters where multiples things can be selected, right?
> *

This is applicable not only to profiles, the submenu with checkboxes always remains open after checking/unchecking.
I agree that the submenu can be closed - when only one option can be selected and when the upper menu shows the selection in the menu item name. This improvement will require new property `close_menu_after_check` (default false) in the touchmenu widget.

**Comment by poire-z** [comment]
No strong opinion either.
I was going to write: _May be not this one - as it give a light feeling of misalignment:_

image

but somehow, I like it - and it makes the 2 first items stand out as the ones that really _do something_.

The other may give a conflicting feeling with their following icons vs. our star and other following icons/indicators in other menus: 
image

So, still no strong opinion :)

> Besides, the above checkboxes can be replaced with radio buttons (we've discussed this earlier).

Don't remember what we may have concluded - but I would have probably say that our checkbox looks nicer than any radio button - and that checkbox were probably just fine in the menu, even if some of them act like radio buttons.

**Comment by NiLuJe** [comment]
> Don't remember what we may have concluded - but I would have probably say that our checkbox looks nicer than any radio button - and that checkbox were probably just fine in the menu, even if some of them act like radio buttons.

I remember you saying something to that effect, yeah ;).

**Comment by NiLuJe** [comment]
No strong opinion on the icons thing either, FWI,W but I do prefer it with them on the left, like @poire-z said ;).

**Comment by hius07** [comment]
What about closing the submenu with checkboxes after checking, when only one option can be selected?
Save one tap.

**Comment by poire-z** [comment]
> What about closing the submenu with checkboxes after checking, when only one option can be selected?

In the context of this PR: I don't mind.
Generically for all our menus, I _do_ mind :) ie. in the font menu where we can select only one font: I don't want it to close, I may want to try another font. I think we may be fine with the current behaviour in other contexts too. But I guess it's in the context where it affects rendering, and you can witness the effect on the part of the book visible below the menu. I don't know if we should really have a different behaviour with menu items that don't have any visual effect. (Although I don't know, we already have `keep_menu_open=true` to have 2 kind of behaviours... :)

**Comment by yparitcher** [comment]
Regression:
This breaks dispatcher. 
The `Nothing` option (and others) will not work with this PR.
Dispatcher is not designed to have other code store items in its data.
Creating a `item.settings` table and storing arbitrary stuff in it is going to break.
We need to find a different place for profile specific settings.

**Comment by poire-z** [comment]
@hius07 : there is a little issue with how you implemented left alignment in Button, the tap area is smaller than the width (here, adding a bordersize to show it):

image
image

I guess because you wanted some left padding so the text is not stuck to the left of the framecontainer.

I guess this extra padding should be put/provided into the existing self.padding_h:
```diff
--- a/frontend/ui/widget/button.lua
+++ b/frontend/ui/widget/button.lua
@@ -84,6 +84,10 @@ function Button:init()

     local is_left_aligned = self.align == "left"
     local right_margin = is_left_aligned and (2 * Size.padding.large) or 0
+    right_margin = 0
+    if is_left_aligned then
+        self.padding_h = Size.padding.large
+    end

     if self.text then
         local max_width = self.max_width
@@ -153,7 +157,7 @@ function Button:init()
     if is_left_aligned then
         self.label_container = LeftContainer:new{
             dimen = Geom:new{
-                w = self.width - 4 * Size.padding.large,
+                w = self.width, -- - 4 * Size.padding.large,
                 h = widget_size.h
             },
             self.label_widget,
```

image

(without borders, fine with that? or would you prefer even more padding ?:)
image

image

Is your profile quickmenu the only current user of align=left ?
I wonder if we should have the "users" of left buttons provide `padding_h=Size.padding.large`, or if Button itself should hardcode (or add) it like I did above. (I guess for flexibility, it should be the users, as we may want to have no padding in some layout.

**Comment by poire-z** [comment]
There is something fishy about how Button handles its width.
I remember I fixed many widgets' width & padding & margin years ago, but I must have forgotten this one...

Currently, we provide a width or max_width, and padding/margins/border are then added to it.
Some widgets workaround that by computing some `inner_button_width=my_required_width - my_padding`, and provide `width=inner_button_width` and ensure margin and border=0.
Some don't, and just live with a larger width - and it seems we are ok with that when there is some centering and this goes unnoticed - or we get some extra padding/margin because the resulting container is bigger than the expected width, which could cause smaller line separators like in this screenshot (which could even be considered as looking nicer, but that's not what the code expected :) just a side effect of this "bug"):
image
Ideally, I guess Button should be fixed with this, if we want it to behave like most other widgets:
```diff
--- a/frontend/ui/widget/button.lua
+++ b/frontend/ui/widget/button.lua
@@ -82,12 +82,13 @@ function Button:init()
         self.padding_v = self.padding
     end

-    local is_left_aligned = self.align == "left"
-    local right_margin = is_left_aligned and (2 * Size.padding.large) or 0
+    local added_width = 2*self.padding_h + 2*self.margin + 2*self.bordersize

     if self.text then
-        local max_width = self.max_width
-            and self.max_width - 2*self.padding_h - 2*self.margin - 2*self.bordersize - right_margin or nil
+        local max_width = self.max_width or self.width
+        if max_width then
+            max_width = max_width - added_width
+        end
         self.label_widget = TextWidget:new{
             text = self.text,
             max_width = max_width,
@@ -147,13 +148,13 @@ function Button:init()
     end
     local widget_size = self.label_widget:getSize()
     if self.width == nil then
-        self.width = widget_size.w
+        self.width = widget_size.w + added_width
     end
     -- set FrameContainer content
-    if is_left_aligned then
+    if self.align == "left" then
         self.label_container = LeftContainer:new{
             dimen = Geom:new{
-                w = self.width - 4 * Size.padding.large,
+                w = self.width,
                 h = widget_size.h
             },
             self.label_widget,
```
but we may then have to fix a few other widgets (where the workaround was obvious), but for some when no expectation is obvious, they'll get bigger and may impact the existing layout and we'll have to notice it visually if things look odd.
Then, for the case in my previous post just above, ButtonTable could provide padding_h = 0 when centered (if we allow for the centered text to possibly reach the borders), and padding_h = Size.padding.large when left aligned, if we want some minimal indentation.

@hius07 : what do you think ?

**Comment by hius07** [comment]
Left-aligned button is used in VocabBuilder too.
I agree that reducing the size of the label_widget is incorrect.
If the correct fix of the Button breaks some other layouts, I do not think it is a big issue, the broken layouts will be fixed then.

**Comment by hius07** [comment]
You know my point of view, I love cleaning everywhere.
It can be done step by step, while discovering/reporting broken places.

**Comment by poire-z** [comment]
Update the patch to button.lua above, and with this in buttontable.lua:
```
+                padding_h = btn_entry.align == "left" and Size.padding.large or 0,
+                    -- allow text to take more of the horizontal space if centered
```
We would get this, with separator lines reaching the borders as when it is centered:
image

**Comment by hius07** [comment]
Very nice, I didn't like smaller separator lines.

**Comment by weijiuqiao** [comment]
>Yeah, VocabBuilder is a festival of various kind of buttons :) @weijiuqiao : did you experience some odd stuff when you did build your layouts, like buttons becoming larger than your specified width and moving other things around? 

I think I assumed everything was expected behavior as that was my first time meddling with KOReader's code :) I do remember having a bit of a hard time trying to put things where they are. It might be a mixture of my inexperience and the fishy button behavior?

>If we go with fixing this, I may request you to check and possibly adapt your VocabBuilder.

It's okay, no worries.

**Comment by poire-z** [comment]
Minor changes to vocabbuilder in the first commit of #10230 - hopefully it won't hurt.
